### ISSUE TITLE
Temporarily force logging with backtrace

### ISSUE THREAD

**Issue by akarzim** [TwP/logging#131]
I would like to force temporarily logging with or without backtrace.  What about this implementation ? https://gist.github.com/akarzim/378314ddc9fc08a47e5c

For now I'm using it as-is in my project, what about a pull request ?

``` ruby
# Temporarily log with or without backtrace
#
# Usage:
# require 'logging'
# logger = Logging.logger(STDOUT)
# logger.with_backtrace do
#   logger.error StandardError.new('oups!')
# end
#
module Logging
  class Logger
    # force backtrace logging temporarily
    def with_backtrace(&block)
      self.class.send(:_backtrace_backup, self)
      self.class.send(:_backtrace_on, self)
      yield block
    ensure
      self.class.send(:_backtrace_restore, self)
    end

    # force no backtrace logging temporarily
    def without_backtrace(&block)
      self.class.send(:_backtrace_backup, self)
      self.class.send(:_backtrace_off, self)
      yield block
    ensure
      self.class.send(:_backtrace_restore, self)
    end

    def self._backtrace_backup(logger)
      @_appenders ||= {}

      logger.appenders.each_with_object(@_appenders) do |appender, accu|
        accu[appender.object_id] = appender.layout.instance_variable_get(:@backtrace)
      end

      _backtrace_backup(logger.parent) if logger.respond_to? :parent
    end
    private_class_method :_backtrace_backup

    def self._backtrace_restore(logger)
      appenders = logger.instance_variable_get(:@appenders)
      logger.appenders.clear

      appenders.each do |appender|
        appender.layout.instance_variable_set(:@backtrace, @_appenders[appender.object_id])
      end

      _backtrace_restore(logger.parent) if logger.respond_to? :parent
    end
    private_class_method :_backtrace_restore

    def self._backtrace_off(logger)
      logger.instance_variable_get(:@appenders).each do |appender|
        appender.layout.instance_variable_set(:@backtrace, false)
      end

      _backtrace_off(logger.parent) if logger.respond_to? :parent
    end
    private_class_method :_backtrace_off

    def self._backtrace_on(logger)
      logger.instance_variable_get(:@appenders).each do |appender|
        appender.layout.instance_variable_set(:@backtrace, true)
      end

      _backtrace_on(logger.parent) if logger.respond_to? :parent
    end
    private_class_method :_backtrace_on
  end
end
```

**Comment by TwP** [comment]
Short answer: this should be done at the [`Layout`](https://github.com/TwP/logging/blob/master/lib/logging/layout.rb#L44)

Longer answer. The `@backtrace` boolean on the layout should have a corresponding setter and accessor so it can be dynamically changed at runtime. However, this setting will affect every message formatted by the layout - i.e. it will apply to all loggers using the layout.

Generally you either want this value turned on at all times or disabled at all times.

**Comment by TwP** [comment]
I just merged in PR #134 that adds accessor methods for the `backtrace` flag on layouts. This should make the code a little bit cleaner.
